File: 4.2.0-alpha02/com.android.tools/sdk-common/com/android/ide/common/gradle/model/level2/IdeModuleLibrary.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model.level2;

import static com.android.ide.common.gradle.model.level2.IdeLibraryFactory.defaultValueIfNotPresent;

import com.android.annotations.NonNull;
import com.android.annotations.Nullable;
import com.android.builder.model.AndroidLibrary;
import com.android.builder.model.JavaLibrary;
import com.android.builder.model.level2.Library;
import com.android.ide.common.gradle.model.IdeModel;
import com.google.common.annotations.VisibleForTesting;
import java.io.File;
import java.util.Collection;
import java.util.Objects;

/** Creates a deep copy of {@link Library} of type LIBRARY_MODULE. */
public final class IdeModuleLibrary implements IdeLibrary {
    @NonNull private final String myArtifactAddress;
    @Nullable private final String myBuildId;
    @Nullable private final String myProjectPath;
    @Nullable private final String myVariant;

    private final boolean myIsProvided;
    private final int myType;
    private final int myHashCode;

    // Used for serialization by the IDE.
    @SuppressWarnings("unused")
    public IdeModuleLibrary() {
        myArtifactAddress = "";
        myBuildId = null;
        myProjectPath = null;
        myVariant = null;
        myIsProvided = false;
        myType = 0;

        myHashCode = 0;
    }

    public IdeModuleLibrary(@NonNull AndroidLibrary library, @NonNull String artifactAddress) {
        myType = LIBRARY_MODULE;
        myArtifactAddress = artifactAddress;
        myBuildId = IdeModel.copyNewProperty(library::getBuildId, null);
        myProjectPath = IdeModel.copyNewProperty(library::getProject, null);
        myVariant = IdeModel.copyNewProperty(library::getProjectVariant, null);
        myIsProvided = defaultValueIfNotPresent(() -> library.isProvided(), false);
        myHashCode = calculateHashCode();
    }

    IdeModuleLibrary(@NonNull JavaLibrary library, @NonNull String artifactAddress) {
        myType = LIBRARY_MODULE;
        myArtifactAddress = artifactAddress;
        myBuildId = IdeModel.copyNewProperty(library::getBuildId, null);
        myProjectPath = IdeModel.copyNewProperty(library::getProject, null);
        myIsProvided = defaultValueIfNotPresent(() -> library.isProvided(), false);
        myVariant = null;
        myHashCode = calculateHashCode();
    }

    @VisibleForTesting
    public IdeModuleLibrary(
            @NonNull String projectPath,
            @NonNull String artifactAddress,
            @Nullable String buildId) {
        myType = LIBRARY_MODULE;
        myArtifactAddress = artifactAddress;
        myBuildId = buildId;
        myProjectPath = projectPath;
        myVariant = null;
        myIsProvided = false;
        myHashCode = calculateHashCode();
    }

    @Override
    public int getType() {
        return myType;
    }

    @Override
    @NonNull
    public String getArtifactAddress() {
        return myArtifactAddress;
    }

    @Override
    @NonNull
    public File getArtifact() {
        throw unsupportedMethodForModuleLibrary("getArtifact()");
    }

    @Nullable
    @Override
    public String getBuildId() {
        return myBuildId;
    }

    @Override
    @Nullable
    public String getProjectPath() {
        return myProjectPath;
    }

    @Override
    @Nullable
    public String getVariant() {
        return myVariant;
    }

    @Override
    @NonNull
    public File getFolder() {
        throw unsupportedMethodForModuleLibrary("getFolder");
    }

    @Override
    @NonNull
    public String getManifest() {
        throw unsupportedMethodForModuleLibrary("getManifest");
    }

    @Override
    @NonNull
    public String getJarFile() {
        throw unsupportedMethodForModuleLibrary("getJarFile");
    }

    @Override
    @NonNull
    public String getCompileJarFile() {
        throw unsupportedMethodForModuleLibrary("getCompileJarFile");
    }

    @Override
    @NonNull
    public String getResFolder() {
        throw unsupportedMethodForModuleLibrary("getResFolder");
    }

    @Nullable
    @Override
    public File getResStaticLibrary() {
        throw unsupportedMethodForModuleLibrary("getResStaticLibrary");
    }

    @Override
    @NonNull
    public String getAssetsFolder() {
        throw unsupportedMethodForModuleLibrary("getAssetsFolder");
    }

    @Override
    @NonNull
    public Collection<String> getLocalJars() {
        throw unsupportedMethodForModuleLibrary("getLocalJars");
    }

    @Override
    @NonNull
    public String getJniFolder() {
        throw unsupportedMethodForModuleLibrary("getJniFolder");
    }

    @Override
    @NonNull
    public String getAidlFolder() {
        throw unsupportedMethodForModuleLibrary("getAidlFolder");
    }

    @Override
    @NonNull
    public String getRenderscriptFolder() {
        throw unsupportedMethodForModuleLibrary("getRenderscriptFolder");
    }

    @Override
    @NonNull
    public String getProguardRules() {
        throw unsupportedMethodForModuleLibrary("getProguardRules");
    }

    @Override
    @NonNull
    public String getLintJar() {
        throw unsupportedMethodForModuleLibrary("getLintJar");
    }

    @Override
    @NonNull
    public String getExternalAnnotations() {
        throw unsupportedMethodForModuleLibrary("getExternalAnnotations");
    }

    @Override
    @NonNull
    public String getPublicResources() {
        throw unsupportedMethodForModuleLibrary("getPublicResources");
    }

    @Override
    @NonNull
    public String getSymbolFile() {
        throw unsupportedMethodForModuleLibrary("getSymbolFile");
    }

    @Override
    public boolean isProvided() {
        return myIsProvided;
    }

    @NonNull
    private static UnsupportedOperationException unsupportedMethodForModuleLibrary(
            @NonNull String methodName) {
        return new UnsupportedOperationException(
                methodName + "() cannot be called when getType() returns LIBRARY_MODULE");
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof IdeModuleLibrary)) {
            return false;
        }
        IdeModuleLibrary that = (IdeModuleLibrary) o;
        return myType == that.myType
                && Objects.equals(myArtifactAddress, that.myArtifactAddress)
                && Objects.equals(myProjectPath, that.myProjectPath)
                && Objects.equals(myBuildId, that.myBuildId)
                && Objects.equals(myVariant, that.myVariant)
                && Objects.equals(myIsProvided, that.myIsProvided);
    }

    @Override
    public int hashCode() {
        return myHashCode;
    }

    private int calculateHashCode() {
        return Objects.hash(
                myType, myArtifactAddress, myBuildId, myProjectPath, myVariant, myIsProvided);
    }

    @Override
    public String toString() {
        return "IdeModuleLibrary{"
                + "myType="
                + myType
                + ", myArtifactAddress='"
                + myArtifactAddress
                + '\''
                + ", myBuildId='"
                + myBuildId
                + '\''
                + ", myProjectPath='"
                + myProjectPath
                + '\''
                + ", myVariant='"
                + myVariant
                + '\''
                + ", myIsProvided='"
                + myIsProvided
                + '\''
                + '}';
    }
}

File: com.android.tools.build/4.2.0-alpha09/com.android.tools/sdk-common/27.2.0-alpha09/com/android/ide/common/gradle/model/impl/IdeModuleLibrary.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model.impl

import com.android.ide.common.gradle.model.IdeLibrary
import com.google.common.annotations.VisibleForTesting
import java.io.File

/**
 * The implementation of IdeLibrary for modules.
 **/
data class IdeModuleLibrary(
  val core: IdeModuleLibraryCore,
  override val isProvided: Boolean
) : IdeLibrary by core {
  @VisibleForTesting
  constructor(
    projectPath: String,
    artifactAddress: String,
    buildId: String?
  ) : this(IdeModuleLibraryCore(projectPath, artifactAddress, buildId), isProvided = false)
}

data class IdeModuleLibraryCore(
  override val artifactAddress: String,
  override val buildId: String?,
  override val projectPath: String?,
  override val variant: String?,
  override val folder: File?,
  override val lintJar: String?

) : IdeLibrary {

  // Used for serialization by the IDE.
  constructor() : this(
    artifactAddress = "",
    buildId = null,
    projectPath = null,
    variant = null,
    folder = null,
    lintJar = null
  )

  constructor(
    projectPath: String,
    artifactAddress: String,
    buildId: String?
  ) : this(
    artifactAddress = artifactAddress,
    buildId = buildId,
    projectPath = projectPath,
    variant = null,
    folder = null,
    lintJar = null
  )

  override val type: IdeLibrary.LibraryType
    get() = IdeLibrary.LibraryType.LIBRARY_MODULE

  override val artifact: File
    get() = throw unsupportedMethodForModuleLibrary("getArtifact()")

  override val manifest: String
    get() = throw unsupportedMethodForModuleLibrary("getManifest")

  override val jarFile: String
    get() = throw unsupportedMethodForModuleLibrary("getJarFile")

  override val compileJarFile: String
    get() = throw unsupportedMethodForModuleLibrary("getCompileJarFile")

  override val resFolder: String
    get() = throw unsupportedMethodForModuleLibrary("getResFolder")

  override val resStaticLibrary: File?
    get() = throw unsupportedMethodForModuleLibrary("getResStaticLibrary")

  override val assetsFolder: String
    get() = throw unsupportedMethodForModuleLibrary("getAssetsFolder")

  override val localJars: Collection<String>
    get() = throw unsupportedMethodForModuleLibrary("getLocalJars")

  override val jniFolder: String
    get() = throw unsupportedMethodForModuleLibrary("getJniFolder")

  override val aidlFolder: String
    get() = throw unsupportedMethodForModuleLibrary("getAidlFolder")

  override val renderscriptFolder: String
    get() = throw unsupportedMethodForModuleLibrary("getRenderscriptFolder")

  override val proguardRules: String
    get() = throw unsupportedMethodForModuleLibrary("getProguardRules")

  override val externalAnnotations: String
    get() = throw unsupportedMethodForModuleLibrary("getExternalAnnotations")

  override val publicResources: String
    get() = throw unsupportedMethodForModuleLibrary("getPublicResources")

  override val symbolFile: String
    get() = throw unsupportedMethodForModuleLibrary("getSymbolFile")

  override val isProvided: Nothing
    get() = error("abstract")
}

private fun unsupportedMethodForModuleLibrary(methodName: String): UnsupportedOperationException {
  return UnsupportedOperationException("$methodName() cannot be called when getType() returns LIBRARY_MODULE")
}
